### PR TITLE
Android Embedding PR24: Allow FlutterActivity to provide an engine, also adjust FlutterFragment timing to avoid Activity launch lag.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Android Embedding PR24: Allow FlutterActivity to provide an engine, also adjust FlutterFragment timing to avoid Activity launch lag.

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java [hunk 1]
```diff
@@ -46,6 +47,8 @@ public void surfaceCreated(SurfaceHolder holder) {
       Log.d(TAG, "SurfaceHolder.Callback.surfaceCreated()");
       isSurfaceAvailableForRendering = true;
 
+      holder.setFormat(PixelFormat.RGBA_8888);
```
Comment (tvolkert) [inline shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java]: Seems like this should be in a separate PR since it's unrelated to the other changes.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 2]
```diff
@@ -108,16 +117,35 @@ public IntentBuilder initialRoute(@NonNull String initialRoute) {
       return this;
     }
 
+    /**
+     * Creates and returns an {@link Intent} that will launch a {@code FlutterActivity} with
+     * the desired configuration.
+     *
+     * Subclasses should override this method to apply additional {@link Intent} arguments.
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Done

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 3]
```diff
@@ -108,16 +117,35 @@ public IntentBuilder initialRoute(@NonNull String initialRoute) {
...
+
+    /**
+     * Returns the specific {@code FlutterActivity} class that should be launched
+     * by this {@code IntentBuilder}.
+     *
+     * {@code FlutterActivity} subclasses should subclass {@code IntentBuilder} and then override
+     * this method to return their specific {@code FlutterActivity} subclass.
+     */
+    @NonNull
+    protected Class<? extends FlutterActivity> getActivityClass() {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 4]
```diff
@@ -258,55 +261,63 @@ private void initializeFlutter(@NonNull Context context) {
   }
 
   /**
-   * Creates a new FlutterEngine instance.
+   * Obtains a reference to a FlutterEngine to back this {@code FlutterFragment}.
    *
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 5]
```diff
@@ -564,4 +575,20 @@ private Context getContextCompat() {
       ? getContext()
       : getActivity();
   }
+
+  /**
+   * Provides a {@link FlutterEngine} instance to be used by a {@code FlutterFragment}.
+   *
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 6]
```diff
@@ -564,4 +575,20 @@ private Context getContextCompat() {
...
+   * {@code FlutterEngineProvider} can be implemented by the {@link FragmentActivity} that
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I tried rewording.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 7]
```diff
@@ -564,4 +575,20 @@ private Context getContextCompat() {
       ? getContext()
       : getActivity();
   }
+
+  /**
+   * Provides a {@link FlutterEngine} instance to be used by a {@code FlutterFragment}.
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I tried expanding the description.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 8]
```diff
@@ -564,4 +575,20 @@ private Context getContextCompat() {
...
+   *
+   * {@code FlutterEngineProvider} can be implemented by the {@link FragmentActivity} that
+   * owns this {@code FlutterFragment}.
+   */
+  public interface FlutterEngineProvider {
+    /**
+     * Returns the {@link FlutterEngine} that should be used by a child {@code FlutterFragment}.
+     *
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Done

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 9]
```diff
@@ -564,4 +575,20 @@ private Context getContextCompat() {
...
+     * This method may return a new {@link FlutterEngine}, or an existing, cached {@link FlutterEngine}.
+     */
+    @Nullable
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Added some clarification for null.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 10]
```diff
@@ -258,55 +261,63 @@ private void initializeFlutter(@NonNull Context context) {
...
-    //                    The existing attach() method does not know how to handle this kind of FlutterView.
-    //flutterEngine.getPluginRegistry().attach(this, getActivity());
-
-    doInitialFlutterViewRun();
+    // We post() the code that attaches the FlutterEngine to our FlutterView because there is
+    // some kind of blocking logic on the native side when the surface is connected. That lag
+    // causes launching Activitys to wait a second or two before launching. By post()'ing this
+    // behavior we are able to move this blocking logic to after the Activity's launch.
+    // TODO(mattcarroll): figure out how to avoid blocking the MAIN thread when connecting a surface
+    new Handler().post(new Runnable() {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I don't think so.  I've never inspected the result of `post()` before, and the only explicit cause of this eventuality from the docs says "usually because the looper processing the message queue is exiting." But we're attaching to the MAIN thread, which won't exit until the application dies. So I think we're OK.

**Comment by tvolkert** [comment]
I don't see any new changes - do you need to push a new commit?

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java]: Removed.

**Comment by matthew-carroll** [comment]
I was marking off as I went. New commit is now pushed.
